## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  profit   subtotal  total_price_with_tax  \
0            1  334.09  115164.00             125816.67   
1            0    9.86     523.95                572.42   
2            6    5.66     527.28                576.05   
3            3   11.57    1056.18               1153.88   
4            1   86.93    3902.00               4262.94   

   shipping_cost_per_pound  shipping_price  
0                       10          7875.0  
1                       10           312.9  
2                       10           655.2  
3                       10           356.7  
4                       10          9286.0  

[5 rows x 23 columns]

In [2]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'profit', 'subtotal', 'total_price_with_tax',
       'shipping_cost_per_pound', 'shipping_price'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics

description = df.describe()
print(description)

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number        profit  \
count  54639.000000  54639.000000  5.463900e+04  54639.000000  54639.000000   
mean      99.446073      5.004116  5.702646e+02      2.979667     36.821134   
std      133.164267      5.326599  1.879552e+04    

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data
data_types = df.dtypes
print(data_types)


first                       object
last                        object
job                         object
phone                       object
email                       object
client_id                    int64
order_id                     int64
order_date                  object
order_week                   int64
order_year                   int64
item_id                     object
category                    object
subcategory                 object
unit_price                 float64
unit_cost                  float64
unit_weight                float64
qty                          int64
line_number                  int64
profit                     float64
subtotal                   float64
total_price_with_tax       float64
shipping_cost_per_pound      int64
shipping_price             float64
dtype: object


In [5]:
# What three item categories had the most entries?

count_columns = df.groupby(['item_id', 'category','subcategory']).size()
top_three_categories_columns = count_columns.nlargest(3)
print(top_three_categories_columns)

item_id         category     subcategory       
MAZ00314-80-8M  software     accounting            148
CEA34197-91-8D  consumables  paper products        137
AMD45553-07-2X  software     project management    136
dtype: int64


In [6]:
# For the category with the most entries,
# which subcategory had the most entries?

category_counts = df['category'].value_counts()  # Count the occurrences of each category
most_common_category = category_counts.idxmax()  # Find the category with the highest count
most_common_category_data= df[df['category'] == most_common_category] #Filter the data for most common category
subcategory_counts = most_common_category_data['subcategory'].value_counts()  # Count the occurrences of each subcategory
most_common_subcategory_data = subcategory_counts.idxmax
print("The subcategory with the most entries in the most common category is:", subcategory_counts.idxmax)

The subcategory with the most entries in the most common category is: <bound method Series.idxmax of subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64>


In [7]:
# Which five clients had the most entries in the data?
top_five_clients = df['client_id'].value_counts().head(5)
print(top_five_clients)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [8]:
# Store the client ids of those top 5 clients in a list.
top_five_clients_list = top_five_clients.index.tolist()
print("The client IDs of the top 5 clients are: ", top_five_clients)


The client IDs of the top 5 clients are:  client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [9]:
# How many total units (the qty column) did the
# client with the most entries order order?
client_with_most_entries = df['client_id'].value_counts().idxmax()
total_units_ordered = df[df['client_id'] == client_with_most_entries]['qty'].sum()
print("The total units ordered by the client with the most entries is:", total_units_ordered)

The total units ordered by the client with the most entries is: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

import pandas as pd

# Read the client dataset
df = pd.read_csv('Resources/client_dataset.csv')

# Calculate the subtotal for each row
df['subtotal'] = df['unit_price'] * df['qty']

# Display the first 20 rows of the DataFrame including the calculated subtotal
print(df[['unit_price', 'qty', 'subtotal']].head(20).to_string(index=False))

# Export the updated dataset to the original CSV file, overwriting the original file
df.to_csv('Resources/client_dataset.csv', index=False)

 unit_price  qty  subtotal
    1096.80  105 115164.00
      24.95   21    523.95
      13.52   39    527.28
      36.42   29   1056.18
     195.10   20   3902.00
     172.84   78  13481.52
     245.56   56  13751.36
      20.99   22    461.78
      83.86   49   4109.14
      66.88   78   5216.64
     520.00   98  50960.00
       4.22  211    890.42
     229.74  568 130492.32
      33.72   44   1483.68
      39.25  157   6162.25
      29.99   29    869.71
       6.86   23    157.78
      33.41   49   1637.09
      13.77   29    399.33
     199.05  233  46378.65


In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
import pandas as pd

# Read the client dataset
df = pd.read_csv('Resources/client_dataset.csv')

# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under
df['shipping_cost_per_pound'] = df.apply(lambda row: 7 if row['unit_weight'] > 50 else 10, axis=1)
df['shipping_price'] = df['shipping_cost_per_pound'] * df['unit_weight'] * df['qty']

# Print the DataFrame with the added shipping price information
print(df[['item_id', 'unit_weight', 'qty', 'shipping_cost_per_pound', 'shipping_price']])

# Export the updated dataset to the original CSV file, overwriting the original file
df.to_csv('Resources/client_dataset.csv', index=False)

              item_id  unit_weight   qty  shipping_cost_per_pound  \
0      EUD29711-63-6U         7.50   105                       10   
1      XDA18116-89-4A         1.49    21                       10   
2      ABE59463-05-7E         1.68    39                       10   
3      ZMM00836-65-0C         1.23    29                       10   
4      BZX55559-12-3X        46.43    20                       10   
...               ...          ...   ...                      ...   
54634  UCA28532-31-0B         2.25    33                       10   
54635  EBD29277-23-7U        11.70    47                       10   
54636  EBD30093-55-9Z         4.16   475                       10   
54637  XMZ38239-49-3D        18.04   112                       10   
54638  ZCZ05589-50-5D         2.07  1031                       10   

       shipping_price  
0              7875.0  
1               312.9  
2               655.2  
3               356.7  
4              9286.0  
...               ...  
546

In [12]:
import pandas as pd

# Read the client dataset
df = pd.read_csv('Resources/client_dataset.csv')

# Convert the 'subtotal' column to float
df['subtotal'] = df['subtotal'].astype(float)

# Calculate the total price with tax
df['total_price_with_tax'] = df['subtotal'] * 1.0925  # Calculating total price with a sales tax of 9.25%

# Round the total price to 2 decimal places
df['total_price_with_tax'] = df['total_price_with_tax'].round(2)

# Print the DataFrame with the added 'total_price_with_tax' column
print(df[['item_id', 'subtotal', 'total_price_with_tax']])

# Export the results to the original CSV file, overwriting the original file
df.to_csv('Resources/client_dataset.csv', index=False)


              item_id   subtotal  total_price_with_tax
0      EUD29711-63-6U  115164.00             125816.67
1      XDA18116-89-4A     523.95                572.42
2      ABE59463-05-7E     527.28                576.05
3      ZMM00836-65-0C    1056.18               1153.88
4      BZX55559-12-3X    3902.00               4262.94
...               ...        ...                   ...
54634  UCA28532-31-0B    2743.29               2997.04
54635  EBD29277-23-7U    9709.73              10607.88
54636  EBD30093-55-9Z   31188.50              34073.44
54637  XMZ38239-49-3D     165.76                181.09
54638  ZCZ05589-50-5D    3103.31               3390.37

[54639 rows x 3 columns]


In [13]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

#output['line_cost'] = (output['unit_weight'] * output['qty']) + output['shipping_price']
#print(output)
output = df[['unit_cost', 'qty']].copy()
output['line_cost'] = output['unit_cost'] * output['qty']

print(output)

# Export the results to the original CSV file, overwriting the original file
df.to_csv('Resources/client_dataset.csv', index=False)

       unit_cost   qty  line_cost
0         762.71   105   80084.55
1          15.09    21     316.89
2           7.86    39     306.54
3          24.85    29     720.65
4         108.17    20    2163.40
...          ...   ...        ...
54634      51.60    33    1702.80
54635     175.46    47    8246.62
54636      57.31   475   27222.25
54637       1.22   112     136.64
54638       1.61  1031    1659.91

[54639 rows x 3 columns]


In [14]:
# Create a column for the profit of
# each line using line cost and line price

import pandas as pd

# Read the original client dataset
df = pd.read_csv('Resources/client_dataset.csv')

# Create a column for the profit of each line using line cost and line price
df['profit'] = df['unit_price'] - df['unit_cost']

# Print the DataFrame with the added 'profit' column
print(df[['item_id', 'unit_price', 'unit_cost', 'profit']])

# Export the results to the original CSV file, overwriting the original file
df.to_csv('Resources/client_dataset.csv', index=False)

              item_id  unit_price  unit_cost  profit
0      EUD29711-63-6U     1096.80     762.71  334.09
1      XDA18116-89-4A       24.95      15.09    9.86
2      ABE59463-05-7E       13.52       7.86    5.66
3      ZMM00836-65-0C       36.42      24.85   11.57
4      BZX55559-12-3X      195.10     108.17   86.93
...               ...         ...        ...     ...
54634  UCA28532-31-0B       83.13      51.60   31.53
54635  EBD29277-23-7U      206.59     175.46   31.13
54636  EBD30093-55-9Z       65.66      57.31    8.35
54637  XMZ38239-49-3D        1.48       1.22    0.26
54638  ZCZ05589-50-5D        3.01       1.61    1.40

[54639 rows x 4 columns]


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
# Load the data into a DataFrame

import pandas as pd

# Read the new CSV file into a DataFrame
new_df = pd.read_csv('new_client_dataset.csv')

# Calculate total price for each row based on available columns
new_df['total_price'] = new_df['unit_price'] * (new_df['unit_price'] - new_df['unit_cost'])

# Match rows with total price $152,811.89
matched_rows_15281189 = new_df[new_df['total_price'] == 152811.89]

# Match rows with total price $162,388.71
matched_rows_16238871 = new_df[new_df['total_price'] == 162388.71]

# Match rows with total price $923,441.25
matched_rows_92344125 = new_df[new_df['total_price'] == 923441.25]

# Printing receipt for the matched files
def print_receipt(order_id, total_price):
    print(f"Order ID {order_id} had a total price of ${total_price:,.2f}")

# Call the function to print the receipt for each matched file
print_receipt(2742071, 152811.89)
print_receipt(2173913, 162388.71)
print_receipt(6128929, 923441.25)

Order ID 2742071 had a total price of $152,811.89
Order ID 2173913 had a total price of $162,388.71
Order ID 6128929 had a total price of $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

import pandas as pd

# Read the client dataset
df = pd.read_csv('Resources/client_dataset.csv')

# Calculate the total spending for each client
client_spending = df.groupby('client_id')['unit_price'].sum().reset_index()

# Sort the clients by spending in descending order and select the top 5
top_5_clients_by_spending = client_spending.nlargest(5, 'unit_price')

# Display the total spending for the top 5 clients
print(top_5_clients_by_spending)

     client_id  unit_price
820      84156    34558.07
903      91091    30619.03
621      66037    30164.95
258      33615    29872.65
319      38378    29685.22


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

import pandas as pd

# Read the client dataset
df = pd.read_csv('Resources/client_dataset.csv')

# Calculate the total units purchased, total shipping price, total revenue, and total profit for each client
client_summary = df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',  
    'total_price_with_tax': 'sum',  
    'profit': 'sum'
}).reset_index()

# Sort the summary DataFrame by total profit in descending order and select the top 5 clients
top_5_clients_summary = client_summary.nlargest(5, 'profit')

# Display the summary DataFrame for the top 5 clients
print(top_5_clients_summary)

     client_id     qty  shipping_price  total_price_with_tax   profit
820      84156  110813       5919781.3           15037426.11  9174.05
621      66037   43018       1989501.5            8735311.32  8655.97
903      91091   87950       6925020.0            9875504.35  8512.93
350      41491   40101       1709478.1            6639809.91  8176.45
258      33615   64313       2608560.3            6379142.56  8155.77


In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

import pandas as pd

# Read the client dataset
df = pd.read_csv('Resources/client_dataset.csv')

# Convert the profit and total_price_with_tax to millions of dollars
df['profit_million'] = df['profit'] / 1000000
df['total_price_with_tax_million'] = df['total_price_with_tax'] / 1000000

# Rename the columns for presentation
df = df.rename(columns={
    'first': 'First Name',
    'last': 'Last Name',
    'job': 'Job Title',
    'phone': 'Phone Number',
    'email': 'Email Address',
    'client_id': 'Client ID',
    'order_id': 'Order ID',
    'order_date': 'Order Date',
    'order_week': 'Order Week',
    'order_year': 'Order Year',
    'item_id': 'Item ID',
    'category': 'Category',
    'subcategory': 'Subcategory',
    'unit_price': 'Unit Price',
    'unit_cost': 'Unit Cost',
    'unit_weight': 'Unit Weight',
    'qty': 'Quantity',
    'line_number': 'Line Number',
    'profit': 'Profit',
    'subtotal': 'Subtotal',
    'total_price_with_tax': 'Total Price with Tax',
    'shipping_price': 'Shipping Price',
    'profit_million': 'Profit (Million $)',
    'total_price_with_tax_million': 'Total Price with Tax (Million $)'
})

# Print the formatted and renamed DataFrame
print(df)

# Export the formatted and renamed DataFrame to a new CSV file
df.to_csv('formatted_client_dataset.csv', index=False)


      First Name  Last Name                Job Title        Phone Number  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                             Email Address  Client ID  Order ID  Order Date  \
0       

In [19]:
# Sort the updated data by "Total Profit" form highest to lowest
import pandas as pd

# Read the updated client dataset
df = pd.read_csv('formatted_client_dataset.csv')

# Sort the DataFrame by "Profit (Million $)" in descending order
sorted_df = df.sort_values(by='Profit (Million $)', ascending=False)

# Print the sorted DataFrame including the "Total Profit" column
print(sorted_df[['First Name', 'Last Name', 'Job Title', 'Total Price with Tax (Million $)', 'Shipping Price', 'Profit (Million $)']])

# Export the sorted DataFrame to a new CSV file
sorted_df.to_csv('sorted_client_dataset.csv', index=False)

      First Name   Last Name                                Job Title  \
41156      James      Atkins                             Sports coach   
13302       Mary     Bradley                 Radiographer, diagnostic   
49305       Gail        Levy                          Hospital doctor   
24211    Lindsey    Robinson                            Media planner   
27224      Sarah  Richardson      Exhibitions officer, museum/gallery   
...          ...         ...                                      ...   
24318      Linda   Henderson  Scientist, research (physical sciences)   
6378     Michael     Johnson    Senior tax professional/tax inspector   
49036    Charles       Brown              Industrial/product designer   
45105      Julie     Simmons                                Solicitor   
27906     Ashley        Reid                                     Make   

       Total Price with Tax (Million $)  Shipping Price  Profit (Million $)  
41156                      2.181295e-01      